### PR TITLE
Add ExpDataSerializer/ExpDataDeserializer

### PR DESCRIPTION
**Pull request by mwjames**

### CODE REVIEW

File: includes/export/SMW_Exp_Data.php [hunk 1]
```diff
@@ -68,6 +68,30 @@ public function getDataItem() {
...
 
 	/**
+	 * @since 2.2
+	 *
+	 * @return string
+	 */
+	public function getHash() {
+
+		$hash = array();
+		$hash[] = $this->getSubject()->getHash();
```
Comment (JeroenDeDauw) [inline includes/export/SMW_Exp_Data.php]: Just access the field directly.

File: includes/export/SMW_Exp_Data.php [hunk 2]
```diff
@@ -68,6 +68,30 @@ public function getDataItem() {
...
+
+		foreach ( $this->getProperties() as $property ) {
```
Comment (JeroenDeDauw) [inline includes/export/SMW_Exp_Data.php]: Same here

File: src/Exporter/Element/ExpElement.php [hunk 3]
```diff
@@ -76,6 +85,28 @@ public function getSerialization() {
 	}
 
 	/**
+	 * @see ExpElement::newFromSerialization
+	 */
+	protected static function deserialize( $serialization ) {
```
Comment (JeroenDeDauw) [inline src/Exporter/Element/ExpElement.php]: Not a fan of having such static deserialization functions, same goes for the `ewFromSerialization` below

File: src/Serializers/ExpDataSerializer.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		if ( !$this->isSerializerFor( $expData ) ) {
+			throw new OutOfBoundsException( 'Object is not supported' );
+		}
+
+		return $this->doSerialize( $expData ) + array( 'serializer' => __CLASS__, 'version' => 0.1 );
+	}
+
+	/**
+	 * @see Serializer::isSerializerFor
+	 *
```
Comment (JeroenDeDauw) [inline src/Serializers/ExpDataSerializer.php]: `isSerializerFor` is not part of `Serializer`, so you do not need to implement it. And if you do implement it, it likely only makes sense for polymorphic usage, in which case you need to implement `DispatchableSerializer`

File: tests/phpunit/includes/Deserializers/ExpDataDeserializerTest.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+			$instance->deserialize( $seralization )
+		);
+
+		$this->assertEquals(
+			$expected->getHash(),
+			$instance->deserialize( $seralization )->getHash()
+		);
+	}
+
+	public function expDataProvider() {
```
Comment (JeroenDeDauw) [inline tests/phpunit/includes/Deserializers/ExpDataDeserializerTest.php]: I highly encourage you to create dedicated test methods for different cases being tested. That makes it a lot easier to understand what is being tested.

**Comment by JeroenDeDauw** [comment]
Introducing these dedicated serializers and deserializers seems good to me. What will they be used for though? Will they replace an existing yet inferior implementation?

File: src/Serializers/ExpDataSerializer.php (hunk 4 above)
Comment (mwjames) [inline src/Serializers/ExpDataSerializer.php]: Noted for a follow-up

File: tests/phpunit/includes/Deserializers/ExpDataDeserializerTest.php (hunk 5 above)
Comment (mwjames) [inline tests/phpunit/includes/Deserializers/ExpDataDeserializerTest.php]: Noted for a follow-up

**Comment by mwjames** [comment]
> Introducing these dedicated serializers and deserializers seems good to me. What will they be used for though? 

You can't send a `ExpData` object into a cache because it could be nested as in `ExpData -> ExpData -> ExpData` etc. and while testing some caching options in the `Exporter` I needed a way the serialize/deserialzer the object from/to cache. The same goes for the hash, now it comes with a stable hasher independent of how child objects where invoked making it actually possible to compare `ExpData` for equality (because the objects are stable and don't rely on any extra `Store` access during a object life-cycle) 

> I'm a bit surprised this has "enhancement" and not "code quality".

Well, it's new and hasn't done before for the `ExpData` object.

**Pull request merged**